Add support for CUE automations

shred86 opened the pull request:
Abode legacy automations and quick actions are being retired on 28 Feb. This pull request adds support for CUE automations which resolves #41. I've tested this along some Home Assistant changes I've made and so far everything is working as expected. I'm having issues with the `tests_automation_enable` method in test_automation.py, but I'll comment directly in that part of the code in this pull request (it's all commented out currently). I'll definitely have more commits as I continue to test and revise the code but I was hoping to start getting some other eyes on.

shred86 commented on tests/test_automation.py:
```diff
@@ -283,109 +266,98 @@ def tests_automation_class_reuse(self, m):
+    #                 aid=AID_1))
+
+    #     # Test the changed state
+    #     automation.enable(False)
```
When `automation.enable(False)` is called, it's raising the `AbodeException((ERROR.INVALID_AUTOMATION_EDIT_RESPONSE)` in automation.py. It's being caused by the `response_object['enabled']` and `self._automation['enabled']` not both being a string or boolean. I'm fairly certain this is caused by using `requests_mock.Mocker()` because I have no issues when I test it live. If I change the automations.py code such that `response_object['enabled']` and `self._automation['enabled']` become `str(response_object['enabled'])` and `str(self._automation['enabled'])`, the test passes fine, but obviously that's probably not a great idea since the whole point is to make sure they're both the same value and type (boolean in this case).

shred86 replied:
After much review and messing around, I don't think there's any reason not to simply it to `str(response_object['enabled'])` and `str(self._automation['enabled'])` (cast both as a string). The most important part is the comparison of whether it's true or false. Whether or not it's a boolean isn't really going to change anything... I could be wrong, but I couldn't think of anything. :)

I did find some other minor issues in the process of looking at this that I resolved as well.

shred86 commented on tests/mock/automation.py:
```diff
@@ -1,15 +1,36 @@
-        "type": "''' + the_type + '''",
-        "notify_msg": null
-  }'''
+        "enabled": "''' + str(enabled) + '''",
```
I did try removing the double quotation marks around `''' + str(enabled) + '''` but that just results in a new error with the json.decoder:

````
Traceback (most recent call last):
  File "/Users/jamiemun/Projects/abodepy/venv/lib/python3.7/site-packages/requests_mock/mocker.py", line 230, in inner
    return func(*args, **kwargs)
  File "/Users/jamiemun/Projects/abodepy/tests/test_automation.py", line 301, in tests_automation_enable
    automation_json = json.loads(automation_text)
  File "/usr/local/Cellar/python/3.7.4_1/Frameworks/Python.framework/Versions/3.7/lib/python3.7/json/__init__.py", line 348, in loads
    return _default_decoder.decode(s)
  File "/usr/local/Cellar/python/3.7.4_1/Frameworks/Python.framework/Versions/3.7/lib/python3.7/json/decoder.py", line 337, in decode
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
  File "/usr/local/Cellar/python/3.7.4_1/Frameworks/Python.framework/Versions/3.7/lib/python3.7/json/decoder.py", line 355, in raw_decode
    raise JSONDecodeError("Expecting value", s, err.value) from None
json.decoder.JSONDecodeError: Expecting value: line 3 column 20 (char 61)
````

shred86 commented:
Coverage decreased because I deleted a test that wasn't needed anymore since quick actions will no longer be a thing (everything will become a CUE automation, to include manually triggered which is the replacement for quick actions). This part of abodepy __init__.py in the `get_automations` method is what's not being covered.

````
if generic_type:
            automations = []
            for automation in self._automations.values():
                if (automation.generic_type is not None and
                        automation.generic_type in generic_type):
                    automations.append(automation)
            return automations
````

I'm not sure if this is even need anymore?

MisterWil commented:
> Coverage decreased because I deleted a test that wasn't needed anymore since quick actions will no longer be a thing (everything will become a CUE automation, to include manually triggered which is the replacement for quick actions). This part of abodepy **init**.py in the `get_automations` method that's not being covered.
> 
> ```
> if generic_type:
>             automations = []
>             for automation in self._automations.values():
>                 if (automation.generic_type is not None and
>                         automation.generic_type in generic_type):
>                     automations.append(automation)
>             return automations
> ```
> 
> I'm not sure if this is even needed anymore since quick actions will be removed.

Well given that you removed the concept of a generic_type from the automations then you should go ahead and remove it from init.py. :-)

Otherwise everything looks good. Since both legacy automations and quick actions are going away anything that references them should be removed from the code.

shred86 commented:
Removed the portion of the code mentioned above. Also removed the `generic_type` and `sub_type` properties from the automation class. The reason being for `generic_type` is because there's obviously only just automations now and the `sub_type` was returning a field that is an empty string in all the JSON data I've looked at for CUE automations (key is subType).

shred86 commented:
Not sure what I changed caused the coverage to decrease. The stuff being highlighted as changed is in __init__.py for abodepy in the `Abode` class. Specifically:

```
    def refresh(self):
        """Do a full refresh of all devices and automations."""
        self.get_devices(refresh=True)
        self.get_automations(refresh=True)
```
Doesn't make any sense to me. This should be triggered based on both test_device.py and test_automation.py. There are test cases where a device or automation id and `refresh=True` being passed to `get_device()` and `get_automation()` which should call `refresh()`.


```
@property
    def uuid(self):
        """Get the UUID."""
        return self._cache[CONST.UUID]
```
Not sure if this was tested in the past but I'm not even sure what this property is used for. I don't see it being called by anything in abodepy. I'm sure it's there for a reason, just can't figure it out.


```
if version and version.lower().startswith('minipir'):
        device_json['generic_type'] = CONST.TYPE_OCCUPANCY
```
This is within the `_new_sensor` method. I don't think it was ever tested before. I guess we just need to add another mocked device (device_type.pov I believe). I have one of these sensors so I can probably update the test for this at some point.
